File: 06ros/ws02/third_ws/src/use_cus_action_01_pkg/src/server_simple_action.cpp
//
// Created by zcb on 5/4/21.
//

#include <actionlib/server/simple_action_server.h>
#include <custom_action_msgs/CountNumberAction.h>
#include <ros/ros.h>

void callback(const custom_action_msgs::CountNumberGoalConstPtr &goal,
              actionlib::SimpleActionServer<custom_action_msgs::CountNumberAction> *server) {
    // 1.实时将进度反馈给client
    // 2.如果执行完毕反馈给client
    // 3.如果client取消，就结束执行
    // 4.server 自己能力有限，干不完，自己abort
    int max = goal->max;
    double duration = goal->duration;

    int cnt = 0; // 从0 开始数数 ,数到 max
    ros::Duration dur(duration);
    bool isCancel = false;
//    bool isAborted = false;
    while (ros::ok() && cnt < max) {
        // 3 检测client 是否取消
        if (server->isPreemptRequested()) {
            isCancel = true;
            break;
        }
        // 4 server 自己崩了
//        if (cnt > 20) {
//            isAborted = true; // 一般情况下,server端 点击 之后 设置isAbort 为 true
//            break;
//        }

        cnt++;
        ROS_INFO_STREAM(cnt);

        //1. 反馈进度
        custom_action_msgs::CountNumberFeedback feedback;
        feedback.rate = cnt / (max * 1.0);
        server->publishFeedback(feedback);


        dur.sleep();
    }
    //2. 反馈结果
    custom_action_msgs::CountNumberResult result;
    result.result = cnt;
    if (isCancel) {
        server->setPreempted(result);
        isCancel = false;
//    } else if (isAborted) {
//        server->setAborted(result);
//        isAborted = false;
    } else {
        server->setSucceeded(result);
    }


}

int main(int argc, char **argv) {
    std::string nodeName = "server_node";
    ros::init(argc, argv, nodeName);

    ros::NodeHandle node;

    std::string actionName = "/zcb01/action";
    actionlib::SimpleActionServer<custom_action_msgs::CountNumberAction> server(
            node, actionName, bind(callback, _1, &server), false);
    server.start();

    ros::spin();


    return 0;
}

File: 06ros/ws02/third_ws/src/use_cus_action_01_pkg/scripts/server_simple_action.py
#! /usr/bin/env python
# coding=utf8

import rospy
from actionlib import SimpleActionServer
from custom_action_msgs.msg import (
    CountNumberAction, CountNumberGoal, CountNumberFeedback,
    CountNumberResult
)

server = None


def callback(goal):
    # 1.实时将进度反馈给client
    # 2.如果执行完毕反馈给client
    # 3.如果client取消，就结束执行
    # 4.server 自己能力有限，干不完，自己abort
    if not isinstance(goal, CountNumberGoal):
        return
    max = goal.max
    dur = goal.duration

    cnt = 0
    duration = rospy.Duration(dur)
    isCancel = False
    # isAborted = False
    while not rospy.is_shutdown() and cnt < max:
        # 3 client 取消
        if server.is_preempt_requested():
            isCancel = True
            break
        # if cnt > 20:
        #     isAborted = True
        #     break

        cnt += 1
        rospy.loginfo(cnt)
        # 1 进度反馈
        feedback = CountNumberFeedback()
        feedback.rate = cnt * 1.0 / max
        server.publish_feedback(feedback)

        rospy.sleep(duration)

    # 2 结果返回
    result = CountNumberResult()
    result.result = cnt
    if isCancel:
        server.set_preempted(result)
        isCancel = False
    # elif isAborted:
    #     server.set_aborted(result)
    #     isAborted = False
    else:
        server.set_succeeded(result)


if __name__ == '__main__':
    nodeName = "server_node"
    rospy.init_node(nodeName)

    actionName = "/zcb01/action"
    server = SimpleActionServer(actionName, CountNumberAction, callback, False)
    server.start()

    rospy.spin()
